### ISSUE TITLE
[Android] Layout Change Causes Control To Be Pushed

### ISSUE THREAD

**Issue by Soap-141** [unoplatform/uno#12315]
### Current behavior

On Android when a UI control in our application is collapsed and then made visible again, it causes a lower control to be displaced and obscured beneath the page's navigation menu.

![Animation_96](https://github.com/unoplatform/uno/assets/57812285/2b73532e-504e-4789-9a93-8f0cbd6795d0)

### Expected behavior

Should work like on UWP/WinUI and iOS.

![Animation_97](https://github.com/unoplatform/uno/assets/57812285/9ed4623f-b95b-45d0-8437-addae5406e70)

### How to reproduce it (as minimally and precisely as possible)

1. Download [FabIssue.zip](https://github.com/unoplatform/uno/files/11490450/FabIssue.zip)
2. Open the application on Android
3. Click multiple times on "Dynamic Control" button
4. Notice the weird behavior of the bottom buttons

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia, Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

4.9.0-dev.1113

### Affected platforms

Android

### IDE

Visual Studio 2022

### IDE version

17.5.5

**Comment by Xiaoy312** [comment]
sample layout:
```cs
Grid // Rows=*,Auto
    0: Grid // Rows=Auto,*
        0: CommandBar
        1: Grid // Rows=Auto,*
            0: StackPanel
                Button // Text='Dynamic Control' Click={ExtendedPart.Visibility = !ExtendedPart.Visibility}
                Grid > TextBlock#ExtendedPart // Text=Extended
            1: ListView#ListViewData // ItemsSource={Enumerable.Range(0,100).Select(x => x.ToString()).ToList()}
            1: [Left,Center,Right].Select(
                (x, i) => Button // Content=Button{i} HV-Align={x},Bottom
            )
    1: Grid > TextBlock // Text=BottomNavigation
```

ExtendedPart.Visibility: // \d=step number, V=Visible, C=Collapsed
- 0V: OK // starting value
- 1C: OK
- 2V: OK
- 3C: OK-ish? // the entire visual of the buttons/page slightly changed (font, shadow, anti-alias)
- 4V: FAIL // buttons got pushed

visual-tree comparison between the steps 0-4: // red-vs-green
> github apparently doesnt support indented diff snippet...

\- 0V-vs-2V: are identical
\- 1C-vs-3C: single/sub-pixel changes on:
```diff
Grid // Rect=[1080x1731@0,0], Actual=411.42857142857144x659.4285714285714, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
-    Grid // Rect=[1080x1616@0,0], Actual=411.42857142857144x615.6190476190476, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
+    Grid // Rect=[1080x1616@0,0], Actual=411.42857142857144x615.4285714285714, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
        CommandBar...
-        Grid // Rect=[1080x1490@0,126], Actual=411.42857142857144x567.6190476190476, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
+        Grid // Rect=[1080x1490@0,126], Actual=411.42857142857144x567.4285714285714, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
            StackPanel...
-            ListView#ListViewData // Rect=[1080x1258@0,232], Actual=411.42857142857144x479.23809523809524, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=[0,0,0,64], Opacity=1, Visibility=Visible
+            ListView#ListViewData // Rect=[1080x1258@0,232], Actual=411.42857142857144x479.04761904761904, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=[0,0,0,64], Opacity=1, Visibility=Visible
-            Button // Rect=[201x115@0,1354], Actual=76.57142857142857x44, HV=Left/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+            Button // Rect=[201x116@0,1353], Actual=76.57142857142857x44, HV=Left/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
-            Button // Rect=[201x115@440,1354], Actual=76.57142857142857x44, HV=Center/Bottom, CornerRadius=4, Margin=[16,0,16,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+            Button // Rect=[201x116@440,1353], Actual=76.57142857142857x44, HV=Center/Bottom, CornerRadius=4, Margin=[16,0,16,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
-            Button // Rect=[201x115@879,1354], Actual=76.57142857142857x44, HV=Right/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+            Button // Rect=[201x116@879,1353], Actual=76.57142857142857x44, HV=Right/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
```
\- 2V-vs-4V:
```diff
while Grid>Grid>Grid and Grid>Grid>Grid>StackPanel stayed the same
the Grid>Grid>Grid>ListView changed size:
    Grid // Rect=[1080x1490@0,126], Actual=411.42857142857144x567.6190476190476, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
        StackPanel // Rect=[1080x348@0,0], Actual=411.42857142857144x132.57142857142856, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=0, Opacity=1, Visibility=Visible
-        ListView#ListViewData // Rect=[1080x1142@0,348], Actual=411.42857142857144x435.04761904761904, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=[0,0,0,64], Opacity=1, Visibility=Visible
+        ListView#ListViewData // Rect=[1080x1258@0,348], Actual=411.42857142857144x479.23809523809524, HV=Stretch/Stretch, CornerRadius=0, Margin=0, Padding=[0,0,0,64], Opacity=1, Visibility=Visible
in addition, the buttons got shifted downward as well
-        Button // Rect=[201x115@0,1354], Actual=76.57142857142857x44, HV=Left/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+        Button // Rect=[201x115@0,1470], Actual=76.57142857142857x44, HV=Left/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
-        Button // Rect=[201x115@440,1354], Actual=76.57142857142857x44, HV=Center/Bottom, CornerRadius=4, Margin=[16,0,16,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+        Button // Rect=[201x115@440,1470], Actual=76.57142857142857x44, HV=Center/Bottom, CornerRadius=4, Margin=[16,0,16,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
-        Button // Rect=[201x115@879,1354], Actual=76.57142857142857x44, HV=Right/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
+        Button // Rect=[201x115@879,1470], Actual=76.57142857142857x44, HV=Right/Bottom, CornerRadius=4, Margin=[0,0,0,8], Padding=[11,5,11,6], Opacity=1, Visibility=Visible, VisualStates=Normal
```
  ^ 4V is quite strange, the LV got shifted down ward which is okay, 
      but retained the size from the previous step (3C) which is NOT correct
      presumably it received a arrange pass, but didnt went through the measure pass...
	  same for the buttons
  ^ the space occupied by children is also bigger than the containing grid (Grid>Grid>Grid):
      [1080x1490@0,126] vs ( [1080x1598@0,0]: [1080x348@0,0](StackPanel) U [1080x1258@0,348](ListView) )
	  ![image](https://github.com/unoplatform/uno/assets/2359550/4dc996bd-e713-46a2-894c-83198eec5708)

**Comment by Xiaoy312** [comment]
tree-graphs of steps 0V~4V: [12315.tree.txt](https://github.com/unoplatform/uno/files/11620525/12315.tree.txt)
best viewed with: https://github.com/Xiaoy312/bag-o-scripts/blob/main/temp/20230419%20text%20comparer.linq

**Comment by Xiaoy312** [comment]
src: https://raw.githubusercontent.com/aosp-mirror/platform_frameworks_base/platform-tools-30.0.0/core/java/android/view/View.java
```java
public final void measure(int widthMeasureSpec, int heightMeasureSpec) {
	...
	if (forceLayout || needsLayout) {
		...
		int cacheIndex = forceLayout ? -1 : mMeasureCache.indexOfKey(key);
		if (cacheIndex < 0 || sIgnoreMeasureCache) {
			// measure ourselves, this should set the measured dimension flag back
			onMeasure(widthMeasureSpec, heightMeasureSpec);
			mPrivateFlags3 &= ~PFLAG3_MEASURE_NEEDED_BEFORE_LAYOUT;
		} else {
			long value = mMeasureCache.valueAt(cacheIndex);
			// Casting a long to int drops the high 32 bits, no mask needed
			setMeasuredDimensionRaw((int) (value >> 32), (int) value);
			mPrivateFlags3 |= PFLAG3_MEASURE_NEEDED_BEFORE_LAYOUT;
		}
		...
	}
	...
}
```

The measure pass inside Grid were ignored due the caching mechanism present in android. The fact that, when the issue finally occurs after a couple visibility toggles, the OnMeasure is called from the arrange(layout) pass confirms this:
```
65 Windows.UI.Xaml.Controls.Panel.OnLayoutCore at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Panel\Panel.Android.cs:79,4
66 Windows.UI.Xaml.FrameworkElement.OnLayoutCore at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\FrameworkElement.Android.cs:254,6
67 Windows.UI.Xaml.Controls.Layouter.Arrange at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Layouter\Layouter.cs:281,5
68 Windows.UI.Xaml.FrameworkElement.FrameworkElementLayouter.ArrangeOverride at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\FrameworkElement.cs:945,63
69 Windows.UI.Xaml.Controls.Grid.ArrangeOverride at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Grid\Grid.cs:1346,5
70 Windows.UI.Xaml.Controls.Grid.InnerArrangeOverride at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Grid\Grid.cs:1433,6
71 Windows.UI.Xaml.FrameworkElement.ArrangeElement at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\FrameworkElement.cs:333,4
72 Windows.UI.Xaml.FrameworkElement.FrameworkElementLayouter.ArrangeElement at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\FrameworkElement.cs:941,64
73 Windows.UI.Xaml.Controls.Layouter.ArrangeChild at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Layouter\Layouter.cs:540,4
74 Windows.UI.Xaml.Controls.Layouter.ArrangeChildOverride at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\Controls\Layouter\Layouter.Android.cs:55,5
75 Android.Views.ViewGroup.Layout at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net7.0/android-33/mcw/Android.Views.ViewGroup.cs:3369,5
76 Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualVoidMethod at /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:93,6
77 Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod at /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/net7.0/JniEnvironment.g.cs:12320,4
78 Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a
79 Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPII_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:154,5
80 Android.Views.View.n_OnMeasure_II at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net7.0/android-33/mcw/Android.Views.View.cs:18941,4
81 Windows.UI.Xaml.FrameworkElement.OnMeasure at D:\code\uno\framework\uno\src\Uno.UI\UI\Xaml\FrameworkElement.Android.cs:193,3
```

Based on the source, the only way to force clear cache were `requestLayout` and `forceLayout`.

**Comment by Soap-141** [comment]
@Xiaoy312 Good job! And thank you!

**Comment by jeromelaban** [comment]
This issue's fixes have been backported to 4.9.49 and 4.10.25.
